src/Plutarch/Test/QuickCheck.hs

```diff
@@ -0,0 +1,265 @@
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Test.QuickCheck (
+    type IsPLam,
+    type IsLast,
+    type PTT,
+    type TestableFun,
+    type PLamArgs,
+    type PA,
+    type PB,
+    type PC,
+    FromPFunN (..),
+    fromPFun,
+    haskEquiv',
+    haskEquiv,
+    shrinkPLift,
+    arbitraryPLift,
+    PFun (..),
+    pattern PFn,
+    TestableTerm (..),
+    PArbitrary (..),
+    pconstantT,
+    pliftT,
+) where
+
+import Data.Kind (Type)
+import Generics.SOP (All, HPure (hcpure), NP ((:*)), Proxy (Proxy))
+import Plutarch (S, Term, (#), type (:-->))
+import Plutarch.Lift (PLift, PUnsafeLiftDecl (PLifted), plift)
+import Plutarch.Prelude (PBool, PInteger)
+import Plutarch.Show (PShow)
+import Plutarch.Test.QuickCheck.Function (
+    PFun (..),
+    applyPFun,
+    plamFinite,
+    plamTable,
+    pattern PFn,
+ )
+import Plutarch.Test.QuickCheck.Instances (
+    PArbitrary (..),
+    TestableTerm (..),
+    pconstantT,
+    pliftT,
+ )
+import Test.QuickCheck (
+    Arbitrary (..),
+    Gen,
+    Property,
+    Testable (property),
+    forAll,
+ )
+
+-- | @since 2.0.0
+data PTermType
+    = LastPFunction
+    | LastPTerm
+    | PFunction
+    | PTerm
+
+-- | @since 2.0.0
+type family PTT' isplam end :: PTermType where
+    PTT' True True = LastPFunction
+    PTT' False True = LastPTerm
+    PTT' True False = PFunction
+    PTT' False False = PTerm
+
+-- | @since 2.0.0
+type PTT (p :: S -> Type) = PTT' (IsPLam p) (IsLast p)
+
+-- | @since 2.0.0
+type family IsPLam (p :: S -> Type) :: Bool where
+    IsPLam ((a :--> b) :--> c) = True
+    IsPLam _ = False
+
+-- | @since 2.0.0
+type family IsLast (p :: S -> Type) :: Bool where
+    IsLast (a :--> PBool) = True
+    IsLast _ = False
+
+{- | Finds Haskell level TestableTerm equivlance of Plutarch
+     functions. This TypeFamily expects the input Plutarch functions to
+     be returning @PBool@ at the end.
+
+     This is used to find type signatures for @quickCheck@able
+     functions from Plutarch terms like @Term s (a :--> b :--> PBool)@.
+
+ @since 2.0.0
+-}
+type family TestableFun (b :: PTermType) (p :: S -> Type) where
+    TestableFun _ PBool = TestableTerm PBool
+    TestableFun PTerm (a :--> b) = TestableTerm a -> TestableFun (PTT b) b
+    TestableFun LastPTerm (a :--> b) = TestableTerm a -> TestableFun (PTT b) b
+    TestableFun PFunction ((a :--> b) :--> c) = PFun a b -> TestableFun (PTT c) c
+    TestableFun LastPFunction ((a :--> b) :--> c) = PFun a b -> TestableFun (PTT c) c
+
+{- | Convert Plutarch function into testable Haskell function that takes
+     `TestableTerm`. It also converts plutarch function into `PFun`.
+
+ @since 2.0.0
+-}
+class FromPFunN (a :: S -> Type) (b :: S -> Type) (c :: PTermType) where
+    fromPFun' :: Proxy c -> (forall s. Term s (a :--> b)) -> TestableFun c (a :--> b)
+
+-- | @since 2.0.0
+instance
+    forall a b.
```

Comment:
Kind signatures.

Reply:
Changing this to 

```hs
forall  {a :: S -> Type} {b :: S -> Type}.
```

yields

```
src/Plutarch/Test/QuickCheck.hs:112:1: error:
    Inferred type variables are not allowed
    In an instance declaration
    |
112 | instance
    | ^^^^^^^^...
```

Reply:
Oh, it should be `()` not `{}`... didn't know there was a difference there.

Reply:
I always used `()`. I wounder what's the difference.